File: smartadapter/src/main/java/smartadapter/SmartEndlessScrollRecyclerAdapter.java
package smartadapter;

/*
 * Created by Manne Öhlund on 2019-07-27.
 * Copyright (c) All rights reserved.
 */

import android.view.ViewGroup;

import androidx.annotation.LayoutRes;
import androidx.annotation.NonNull;

import java.util.ArrayList;
import java.util.List;

import io.github.manneohlund.smartrecycleradapter.R;
import smartadapter.listener.OnLoadMoreListener;
import smartadapter.viewholder.LoadMoreViewHolder;
import smartadapter.viewholder.SmartViewHolder;

/**
 * Enables endless scrolling or pagination. Let's the adapter show a {@link LoadMoreViewHolder} when scrolled to last item.
 */
public class SmartEndlessScrollRecyclerAdapter extends SmartRecyclerAdapter implements ISmartEndlessScrollRecyclerAdapter {

    private final int VIEW_TYPE_LOADING = Integer.MAX_VALUE;

    private OnLoadMoreListener onLoadMoreListener;
    private boolean endlessScrollEnabled = true;
    private boolean autoLoadMoreEnabled = true;
    private boolean loading = false;
    private int loadMoreLayoutResource = R.layout.load_more_view;

    public SmartEndlessScrollRecyclerAdapter(Object callerEnclosingClass, @NonNull List items) {
        super(callerEnclosingClass, items);
    }

    @Override
    public int getItemViewType(int position) {
        if (isEndlessScrollEnabled() && position == getItemCount() - getEndlessScrollOffset()) {
            return VIEW_TYPE_LOADING;
        }
        return super.getItemViewType(position);
    }

    @NonNull
    @Override
    public SmartViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        if (viewType == VIEW_TYPE_LOADING) {
            return new LoadMoreViewHolder(parent, loadMoreLayoutResource, autoLoadMoreEnabled);
        }
        return super.onCreateViewHolder(parent, viewType);
    }

    @Override
    public void onBindViewHolder(@NonNull SmartViewHolder holder, int position) {
        if (position < getItemCount() - getEndlessScrollOffset()) {
            super.onBindViewHolder(holder, position);
        }
    }

    @Override
    public int getItemCount() {
        return super.getItemCount() + getEndlessScrollOffset();
    }

    @Override
    public void onViewAttachedToWindow(@NonNull SmartViewHolder holder) {
        super.onViewAttachedToWindow(holder);
        if (holder instanceof LoadMoreViewHolder) {
            if (autoLoadMoreEnabled) {
                onLoadMoreListener.onLoadMore((LoadMoreViewHolder) holder);
            } else {
                ((LoadMoreViewHolder)holder).toggleLoading(false);
                holder.itemView.findViewById(R.id.loadMoreButton).setOnClickListener(v -> {
                    onLoadMoreListener.onLoadMore((LoadMoreViewHolder) holder);
                    ((LoadMoreViewHolder)holder).toggleLoading(true);
                });
            }
        }
    }

    @Override
    public int getEndlessScrollOffset() {
        return isEndlessScrollEnabled() ? 1 : 0;
    }

    @Override
    public boolean isEndlessScrollEnabled() {
        return endlessScrollEnabled;
    }

    @Override
    public void setEndlessScrollEnabled(boolean enabled) {
        this.endlessScrollEnabled = enabled;
        smartNotifyItemChanged(getItemCount());
    }

    @Override
    public void setAutoLoadMore(boolean enabled) {
        this.autoLoadMoreEnabled = enabled;
    }

    @Override
    public boolean isLoading() {
        return loading;
    }

    @Override
    public void setIsLoading(boolean loading) {
        this.loading = loading;
    }

    @Override
    public void setOnLoadMoreListener(@NonNull OnLoadMoreListener onLoadMoreListener) {
        this.onLoadMoreListener = onLoadMoreListener;
    }

    @Override
    public void setCustomLoadMoreLayoutResource(@LayoutRes int loadMoreLayoutResource) {
        this.loadMoreLayoutResource = loadMoreLayoutResource;
    }

    /**
     * Builder of {@link SmartRecyclerAdapter} for easy implementation.
     * @return SmartAdapterBuilder
     */
    public static SmartAdapterBuilder items(List items) {
        return new SmartAdapterBuilder(new SmartEndlessScrollRecyclerAdapter(null, items));
    }

    /**
     * Builder of {@link SmartRecyclerAdapter} for easy implementation.
     * @return SmartAdapterBuilder
     */
    public static SmartAdapterBuilder empty() {
        return new SmartAdapterBuilder(new SmartEndlessScrollRecyclerAdapter(null, new ArrayList()));
    }
}

File: smartadapter/src/main/java/smartadapter/SmartEndlessScrollRecyclerAdapter.kt
package smartadapter

/*
 * Created by Manne Öhlund on 2019-07-27.
 * Copyright (c) All rights reserved.
 */

import android.view.View
import android.view.ViewGroup
import androidx.annotation.LayoutRes
import io.github.manneohlund.smartrecycleradapter.R
import smartadapter.internal.extension.isMutable
import smartadapter.listener.OnLoadMoreListener
import smartadapter.viewholder.LoadMoreViewHolder
import smartadapter.viewholder.SmartViewHolder

/**
 * Enables endless scrolling or pagination. Let's the adapter show a [LoadMoreViewHolder] when scrolled to last item.
 */
@Suppress("UNCHECKED_CAST")
class SmartEndlessScrollRecyclerAdapter(items: MutableList<Any>) : SmartRecyclerAdapter(items), ISmartEndlessScrollRecyclerAdapter {

    private val VIEW_TYPE_LOADING = Integer.MAX_VALUE

    override var isEndlessScrollEnabled: Boolean = true
        set(value) {
            field = value
            smartNotifyItemChanged(itemCount)
        }
    override var isLoading: Boolean = false
    override var endlessScrollOffset: Int = if (isEndlessScrollEnabled) 1 else 0
    override var autoLoadMoreEnabled: Boolean = false
    private var onLoadMoreListener: OnLoadMoreListener? = null
    private var loadMoreLayoutResource = R.layout.load_more_view

    override fun getItemViewType(position: Position): ViewType {
        return if (isEndlessScrollEnabled && position == itemCount - endlessScrollOffset) {
            VIEW_TYPE_LOADING
        } else super.getItemViewType(position)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: ViewType): SmartViewHolder<Any> {
        return if (viewType == VIEW_TYPE_LOADING) {
            LoadMoreViewHolder(parent, loadMoreLayoutResource, autoLoadMoreEnabled)
        } else super.onCreateViewHolder(parent, viewType)
    }

    override fun onBindViewHolder(holder: SmartViewHolder<Any>, position: Position) {
        if (position < itemCount - endlessScrollOffset) {
            super.onBindViewHolder(holder, position)
        }
    }

    override fun getItemCount(): Int {
        return super.getItemCount() + endlessScrollOffset
    }

    override fun onViewAttachedToWindow(holder: SmartViewHolder<Any>) {
        super.onViewAttachedToWindow(holder)
        if (holder is LoadMoreViewHolder) {
            if (autoLoadMoreEnabled) {
                onLoadMoreListener?.invoke(holder)
            } else {
                holder.toggleLoading(false)
                holder.itemView.findViewById<View>(R.id.loadMoreButton).setOnClickListener {
                    onLoadMoreListener?.invoke(holder)
                    holder.toggleLoading(true)
                }
            }
        }
    }

    override fun setOnLoadMoreListener(onLoadMoreListener: OnLoadMoreListener) {
        this.onLoadMoreListener = onLoadMoreListener
    }

    override fun setCustomLoadMoreLayoutResource(@LayoutRes loadMoreLayoutResource: Int) {
        this.loadMoreLayoutResource = loadMoreLayoutResource
    }

    companion object {

        /**
         * Builder of [SmartRecyclerAdapter] for easy implementation.
         * @return SmartAdapterBuilder
         */
        fun items(items: List<*>): SmartAdapterBuilder =
            SmartAdapterBuilder(SmartEndlessScrollRecyclerAdapter(items.let {
                (if (it.isMutable()) it else it.toMutableList()) as MutableList<Any>
            }))

        /**
         * Builder of [SmartRecyclerAdapter] for easy implementation.
         * @return SmartAdapterBuilder
         */
        fun empty(): SmartAdapterBuilder =
            SmartAdapterBuilder(SmartEndlessScrollRecyclerAdapter(mutableListOf()))
    }
}
